### PR TITLE
PLAT-961| fix api includes performance

### PR DESCRIPTION
**Pull request by kvas-damian**
* **singular vs plural in include parameter is important!** - we validate if the relation exists on a model and we miss the relation when we use wrong include
* add `eagerLoadIncludes()` for `.filter` endpoint and endpoints using `nestedSets`
* we have to pass camelCase class names to Eloquent `->with()` method and also our includes validator was using them to check if a relation exists on a model

Results:
* http://platform.local/papi/v2/taxonomy_terms/byTaxonomy/5?include=tag 
```
|               | before | after |
|---------------|--------|-------|
| db queries    | 1305   | 4     |
| response time | 1402ms | 220ms |
```

* http://platform.local/papi/v2/annotations/.filter
```
|               | before | after |
|---------------|--------|-------|
| db queries    | 276    | 8     |
| response time | 334ms  | 74ms  |
```

Example: http://platform.local/papi/v2/taxonomy_terms/byTaxonomy/6?include=tag
```
{  
   "0":{  
      "id":7100,
      "description":"",
      "parent_id":null,
      "tag":[1707]
   },
   "included":{  
      "tags":{  
         "1707":{  
            "id":1707,
            "name":"zzz",
            "description":null,
            "color":null,
            "taxonomy_terms":7100,
            "0":[]
         }
      }
   }
}
```

To test:
* [ ] Content classifier
* [ ] Taxonomy editor
* [ ] Users list
* [ ] Tags list
* [ ] My orders - user facing

### CODE REVIEW

File: app/Http/Controllers/Api/ApiController.php [hunk 1]
```diff
@@ -260,7 +260,8 @@ protected function loadCountIncludes($model) {
 
 	protected function modelHasMethod(string $model, string $method)
 	{
-		return method_exists((new $model), $method);
+		// TODO PLAT-960 remove the hardcoded class and solve the problem of loading includes that arend defined by a simple relation to another class
+		return $model === 'App\\Models\\Ancestor' || method_exists((new $model), $method);
 	}
 
 	protected function processChain(string $chain, string $parentModel)
```
Comment (gitkuba) [inline app/Http/Controllers/Api/ApiController.php]: Change to `\Log::warning` or `\Log::info`.

File: app/Http/Controllers/Api/Transformers/TaxonomyTermTransformer.php [hunk 2]
```diff
@@ -27,14 +27,14 @@ public function transform(TaxonomyTerm $taxonomyTerm) {
 		return $data;
 	}
 
-	public function includeTags(TaxonomyTerm $taxonomyTerm)
+	public function includeTag(TaxonomyTerm $taxonomyTerm)
 	{
-		return $this->item($taxonomyTerm->tag, new TagTransformer(['taxonomy_terms' => $taxonomyTerm->id]), 'tags');
+		return $this->item($taxonomyTerm->tag, new TagTransformer(['taxonomy_terms' => $taxonomyTerm->id]), 'tag');
```
Comment (gitkuba) [inline app/Http/Controllers/Api/Transformers/TaxonomyTermTransformer.php]: Consider staying by plural form for consistency in resource naming.

File: app/Http/Controllers/Api/ApiController.php (hunk 1 above)
Comment (kvas-damian) [reply app/Http/Controllers/Api/ApiController.php]: Done

File: app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php [hunk 3]
```diff
@@ -81,8 +82,15 @@ public function includedData(ResourceInterface $resource, array $data)
...
-					$this->includes[$includedResourceName][$item['id']] = $item;
+					// make name for list of included resources plural
+					$pluralIncludedResourceName = Str::plural($includedResourceName);
+
+					if (!isset($this->relationships[$resourceKey][$resourceId][$pluralIncludedResourceName]) ||
+						!in_array($item['id'], $this->relationships[$resourceKey][$resourceId][$pluralIncludedResourceName], true)) {
+						// Don't duplicate relation id when parent resource was included multiple times
+						$this->relationships[$resourceKey][$resourceId][$pluralIncludedResourceName][] = $item['id'];
+					}
+					$this->includes[$pluralIncludedResourceName][$item['id']] = $item;
```
Comment (rogatty) [inline app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php]: Wawaweewa, how about adding a unit test for this method?

File: app/Http/Controllers/Api/Transformers/TaxonomyTermTransformer.php [hunk 4]
```diff
@@ -7,7 +7,7 @@
 
 class TaxonomyTermTransformer extends ApiTransformer {
 	protected $parent;
-	protected $availableIncludes = ['tags', 'taxonomies', 'ancestors'];
+	protected $availableIncludes = ['tag', 'taxonomy', 'ancestors'];
```
Comment (rogatty) [inline app/Http/Controllers/Api/Transformers/TaxonomyTermTransformer.php]: We wanted to use plurals for resource names, no?

Comment (kvas-damian) [reply app/Http/Controllers/Api/Transformers/TaxonomyTermTransformer.php]: I'm not sure about this one. How do you know then if the relationship is to one or many items?

File: app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php [hunk 5]
```diff
@@ -77,12 +78,20 @@ public function includedData(ResourceInterface $resource, array $data)
...
-					$this->includes[$includedResourceName][$item['id']] = $item;
+
+					if (!isset($this->relationships[$resourceKey][$resourceId][$includedResourceName]) ||
+						!in_array($item['id'], $this->relationships[$resourceKey][$resourceId][$includedResourceName], true)) {
+						// Don't duplicate relation id when parent resource was included multiple times
+						$this->relationships[$resourceKey][$resourceId][$includedResourceName][] = $item['id'];
+					}
+
+					// make name for list of included resources plural
+					$this->includes[Str::plural($includedResourceName)][$item['id']] = $item;
```
Comment (rogatty) [inline app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php]: Removing this pluralization is a part of PLAT-971, yes?

Comment (kvas-damian) [reply app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php]: Nope, pluralization is performed here.

Comment (rogatty) [reply app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php]: But if I want to include `tag` not `tags` then it shouldn't convert it to `tags` in here, yes?

Comment (kvas-damian) [reply app/Http/Controllers/Api/Serializer/ApiJsonSerializer.php]: without this change the output is strange:
`
...
tag: [5]
included: {
  tag: {5: {...}, 10{...}}
}
`

After this change we make the resource name plural in the output:
`
...
tag: [5]
included: {
  tags: {5: {...}, 10{...}}
}
`

File: app/Http/Controllers/Api/ApiController.php [hunk 6]
```diff
@@ -270,14 +271,14 @@ protected function processChain(string $chain, string $parentModel)
 		for ($i = 0; $i < $depth; $i++) {
 			if ($i === 0) {
 				if (!$this->modelHasMethod($parentModel, $resources[$i])) {
-					\Log::debug("Relationship {$resources[$i]} does not exist in model {$parentModel}");
+					\Log::warning("Relationship {$resources[$i]} does not exist in model {$parentModel}");
```
Comment (bognix) [inline app/Http/Controllers/Api/ApiController.php]: warning goes to error-reporter, we want that? \Log::notice will send to sentry but don't throw in error reporter. I think there will be a lot of those errors, but we will see :)

Comment (kvas-damian) [reply app/Http/Controllers/Api/ApiController.php]: @gitkuba changed that recently

File: tests/Unit/Http/Controllers/Api/Serializer/ApiJsonSerializerTest.php [hunk 7]
```diff
@@ -0,0 +1,204 @@
+<?php
+/**
```
Comment (bognix) [inline tests/Unit/Http/Controllers/Api/Serializer/ApiJsonSerializerTest.php]: this comment can be removed :)

**Pull request merged**